Add raycast-ia-writer extension

xvvhang opened the pull request:
## Description

There are two commands in this extension to let users could create or search notes quickly inside iA Writer library.

With Quick Note command, users could jot down anything and specify the filename of their notes or use time stamp as filenames.
With Quick Search command, users could open iA Writer's Quick Search window with query instantly.

The extension is made with iA Writer's URL schemes.

## Screencast

![quick-note](https://github.com/raycast/extensions/assets/46291413/bd64bd56-f737-4735-84d6-900d4491b69c)

![quick-search](https://github.com/raycast/extensions/assets/46291413/27ae7f16-b905-47b5-a3ca-06c7e8af36be)

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

pernielsentikaer reviewed (changes requested):
Hi 👋 

Thanks for your contribution 💪 

I have now tested your extension, and I have some feedback ready for you:

- Maybe you can check if Ai Writer is installed? You can see how extensions like `Eagle` does it 🙂

- Why is current datetime populated in the form, is that something most people wants?

- Could you add at least one [screenshot](https://developers.raycast.com/basics/prepare-an-extension-for-store#screenshots) of the extension

I'm looking forward to testing this extension again 🔥 

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

xvvhang commented:
> Hi 👋
> 
> Thanks for your contribution 💪
> 
> I have now tested your extension, and I have some feedback ready for you:
> 
> * Maybe you can check if Ai Writer is installed? You can see how extensions like `Eagle` does it 🙂
> * Why is current datetime populated in the form, is that something most people wants?
> * Could you add at least one [screenshot](https://developers.raycast.com/basics/prepare-an-extension-for-store#screenshots) of the extension
> 
> I'm looking forward to testing this extension again 🔥
> 
> Request a new review when you are ready. Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

Hi

Thanks for your review. I’ve got your feedback and made some changes:

1. Added installation check like what `Eagle` extension did.
2. Added screenshots to `/metadata` folder.

And I know your confusion about the datetime populated in Quick Note form. It’s a common feature in some note apps like Obsidian([Unique note creator](https://help.obsidian.md/Plugins/Unique+note+creator)), which makes it easy to create quick notes in a [Zettelkasten](https://en.wikipedia.org/wiki/Zettelkasten) way. There are options to disable the feature or specify datetime format. So users can use this extension in their ways.

Looking forward to more feedbacks.
Thank you!

pernielsentikaer commented on extensions/raycast-ia-writer/src/quick-note.tsx:
```diff
@@ -0,0 +1,49 @@
+      }
+    >
+      <Form.TextField id="name" title="Filename" value={name} onChange={(name) => setName(name)} />
+      <Form.TextArea id="text" title="Content" value={text} onChange={(text) => setText(text)} />
```
```suggestion
      <Form.TextArea
        id="text"
        title="Content"
        value={text}
        autoFocus={defaultText() !== ""}
        onChange={(text) => setText(text)}
      />
```
Maybe set autofocus when date is set?

pernielsentikaer reviewed (changes requested):
Hi 👋

I tested the new improvements and have some feedback 🙂

I still see "Note Saved" if I enters the wrong credentials, do you have an idea why?

xvvhang commented:
> Hi 👋
> 
> I tested the new improvements and have some feedback 🙂
> 
> I still see "Noe Saved" if I enters the wrong credentials, do you have an idea why?

Hi  👋

The Quick Note command calls mac's `open` command with iA Writer's URL Scheme in a node child process. The `open` command runs successfully even there's something wrong with URL's query string. So I couldn't know if the note is created or not. 🥲

I checked how `Bear` and `Obsidian` deal with the case. `Bear` just opens the URL without any success toasts. And `Obsidian` creates a new file or overwrites the file if it's existed.

I prefer the `Bear` way and may remove the `Note Saved` toast. If the note's not saved correctly. iA Writer just pop up. What do you think of it ? Should I remove the toast? 🤔

Thanks for your feedback!

pernielsentikaer commented:
Let's do that, this was we don't confuse the end user 🙂

mil3na commented:
Hi!

I'm taking a look at your extension while Per is on vacay :D 
By first running it, without IA Writer installed, I can see an error on the bottom of the screen. However, I can still see the inputs and ⌘Enter "saves the note" successfully. I don't believe this would be the best approach to this. The first screen should ask the user all the things needed for the extension to work properly. 
Do you mind changing that?

Also, the quick search is not working for me. It does nothing other than show a HUD.

xvvhang commented:
> Hi!
> 
> I'm taking a look at your extension while Per is on vacay :D By first running it, without IA Writer installed, I can see an error on the bottom of the screen. However, I can still see the inputs and ⌘Enter "saves the note" successfully. I don't believe this would be the best approach to this. The first screen should ask the user all the things needed for the extension to work properly. Do you mind changing that?
> 
> Also, the quick search is not working for me. It does nothing other than show a HUD.

Hi,

I just removed the wrong toast when iA writer is not installed and updated the icon.
As Per asked, I checked how `Eagle` extension works when Eagle app not installed. It just shows a toast asking users to install Eagle app. And I think this extension does it in the same way. Could you please check how other extensions work when apps not installed like `Obsidian` and `Bear`?

Thanks for your review!

mil3na commented:
Hey! 
I understand that Per suggested it, but I don't think it is a good UI. It is confusing for the user. You could do the same check you are doing now, but show a different "screen" with instructions for the user, in case they don't have the app installed.

xvvhang commented:
> Hey! I understand that Per suggested it, but I don't think it is a good UI. It is confusing for the user. You could do the same check you are doing now, but show a different "screen" with instructions for the user, in case they don't have the app installed.

Hi,

I understand your concern. But I wonder why there isn't a 'different' screen in `Bear` or `Obsidian` extension when I don't have Bear or Obsidian installed? Both of them are published to the store and got thousands of downloads. I don't really understand.

If it's required to add such a 'different' screen, should I use an `Alert` or push to a new `Detail` page? Could you please give me some best practices examples to learn from?

Thanks for your review.

pernielsentikaer commented:
We could do something like this @xvvhang 🙂

![raycast-ia-writer 2024-04-03 at 16 28 29](https://github.com/raycast/extensions/assets/38033107/e0b78eed-ecdc-4b97-b938-5bd4c9eb5f2a)

You're right there is other extensions in the store that doesn't handle it but might - most people installing this extension should have AI Writer so it's not the worst issue - could you check my additions?

xvvhang commented:
> We could do something like this @xvvhang 🙂
> 
> ![raycast-ia-writer 2024-04-03 at 16 28 29](https://private-user-images.githubusercontent.com/38033107/319228240-e0b78eed-ecdc-4b97-b938-5bd4c9eb5f2a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI0NzgyMzIsIm5iZiI6MTcxMjQ3NzkzMiwicGF0aCI6Ii8zODAzMzEwNy8zMTkyMjgyNDAtZTBiNzhlZWQtZWNkYy00Yjk3LWI5MzgtNWJkNGM5ZWI1ZjJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA3VDA4MTg1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkYjhkM2FjYThmOTNlZDY5MTljZDk2NTc0YWExZjA0ODcxMmU1NGJlNjYzMGFkODRhZjQ3ZDIwNTZiN2ViOTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5MaNF2IzW6r_sqJg-x_Udo1tTt_I4CaRb_NuF1P_pb0)
> 
> You're right there is other extensions in the store that doesn't handle it but might - most people installing this extension should have AI Writer so it's not the worst issue - could you check my additions?

Sure! That make senses and I'm glad to make some changes. I will commit some changes later. Thank you Per!🫡

pernielsentikaer commented:
I already made above in the extension, can you approve it? 🙂

xvvhang reviewed (commented):
LGTM!

xvvhang commented:
Thanks for your help! 🙂@pernielsentikaer

pernielsentikaer reviewed (approved):
Hi 👋 

Looks good to me, approved 🔥
